src/deluge/model/drum/kit.cpp

```diff
@@ -994,177 +994,182 @@ void Kit::getThingWithMostReverb(Sound** soundWithMostReverb, ParamManager** par
+					if (getRootUI()) {
+						getRootUI()->noteRowChanged(instrumentClip, thisNoteRow);
 					}
-
-					beginAuditioningforDrum(modelStackWithNoteRow, thisDrum, velocity, mpeValues, channel);
 				}
 			}
+			instrumentClipView.reportNoteOffForMPEEditing(modelStackWithNoteRow);
 
-			// Note-off
-			else {
-				if (thisNoteRow) {
-					if (shouldRecordNotes && thisDrum->auditioned
-					    && ((playbackHandler.recording == RECORDING_ARRANGEMENT
-					         && instrumentClip->isArrangementOnlyClip())
-					        || currentSong->isClipActive(instrumentClip))) {
-
-						if (playbackHandler.recording == RECORDING_ARRANGEMENT
-						    && !instrumentClip->isArrangementOnlyClip()) {}
-						else {
-							instrumentClip->recordNoteOff(modelStackWithNoteRow, velocity);
-							if (getRootUI()) {
-								getRootUI()->noteRowChanged(instrumentClip, thisNoteRow);
-							}
-						}
-					}
-					instrumentClipView.reportNoteOffForMPEEditing(modelStackWithNoteRow);
-
-					// MPE-controlled params are a bit special in that we can see (via this note-off) when the user has removed their finger and won't be sending more
-					// values. So, let's unlatch those params now.
-					ExpressionParamSet* mpeParams = thisNoteRow->paramManager.getExpressionParamSet();
-					if (mpeParams) {
-						mpeParams->cancelAllOverriding();
-					}
-				}
-				endAuditioningForDrum(
-				    modelStackWithNoteRow, thisDrum,
-				    velocity); // Do this even if not marked as auditioned, to avoid stuck notes in cases like if two note-ons were sent
+			// MPE-controlled params are a bit special in that we can see (via this note-off) when the user has removed their finger and won't be sending more
+			// values. So, let's unlatch those params now.
+			ExpressionParamSet* mpeParams = thisNoteRow->paramManager.getExpressionParamSet();
+			if (mpeParams) {
+				mpeParams->cancelAllOverriding();
 			}
 		}
+		endAuditioningForDrum(
+		    modelStackWithNoteRow, thisDrum,
+		    velocity); // Do this even if not marked as auditioned, to avoid stuck notes in cases like if two note-ons were sent
+	}
+}
+
+void Kit::offerReceivedNote(ModelStackWithTimelineCounter* modelStack, MIDIDevice* fromDevice, bool on, int32_t channel,
+                            int32_t note, int32_t velocity, bool shouldRecordNotes, bool* doingMidiThru) {
 
+	InstrumentClip* instrumentClip = (InstrumentClip*)modelStack->getTimelineCounterAllowNull(); // Yup it might be NULL
+
+	for (Drum* thisDrum = firstDrum; thisDrum; thisDrum = thisDrum->next) {
+
+		// If this is the "input" command, to sound / audition the Drum...
+		// Returns true if midi channel and note match the learned midi note
+		// We don't need the MPE match because all types of matches should sound the drum
+		if (thisDrum->midiInput.equalsNoteOrCCAllowMPE(fromDevice, channel, note)) {
+			receivedNoteForDrum(modelStack, fromDevice, on, channel, note, velocity, shouldRecordNotes, doingMidiThru,
+			                    thisDrum);
+		}
```

Comment:
I think there's a behavior change here -- in the original version of the loop, only the first drum that claims the note when `on == true` will get that event. This version doesn't do that, I think.

Reply:
`lookingForFirstDrumForNoteOn` in the original code.

Reply:
Yeah it's intentional - removes a micro optimization to simplify the code. It only saves a check if two drums in the same kit are mapped to the same midi note